src/test/java/space/obminyashka/items_exchange/service/EmailServiceTest.java

```diff
@@ -0,0 +1,96 @@
+import space.obminyashka.items_exchange.dao.EmailConfirmationCodeRepository;
+import space.obminyashka.items_exchange.dao.UserRepository;
+import space.obminyashka.items_exchange.exception.EmailValidationCodeExpiredException;
+import space.obminyashka.items_exchange.exception.EmailValidationCodeNotFoundException;
+import space.obminyashka.items_exchange.model.EmailConfirmationCode;
+import space.obminyashka.items_exchange.model.User;
+import space.obminyashka.items_exchange.service.impl.SendGridService;
+import space.obminyashka.items_exchange.util.MessageSourceUtil;
+
+import java.time.LocalDateTime;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@ExtendWith(MockitoExtension.class)
+public class EmailServiceTest {
+    public static final UUID id = UUID.fromString("61731cc8-8104-49f0-b2c3-5a52e576ab18");
+    @Mock
+    private UserRepository userRepository;
+    @Mock
+    private MessageSource messageSource;
+    @Mock
+    private EmailConfirmationCodeRepository emailConfirmationCodeRepository;
+    @Captor
+    private ArgumentCaptor<UUID> IdArgumentCaptor;
+    @InjectMocks
+    private SendGridService sendGridService;
+    @InjectMocks
+    private MessageSourceUtil messageSourceUtil;
+
+    @BeforeEach
+    void init() {
+        messageSourceUtil.setMSource(messageSource);
+    }
+
+    @Test
+    void TestValidateEmailCode_WhenDataCorrect_Successfully() throws EmailValidationCodeExpiredException, EmailValidationCodeNotFoundException {
+        when(emailConfirmationCodeRepository.findById(id)).thenReturn(creatEmailConfirmCode());
+        sendGridService.validateEmail(id);
+
+        verify(userRepository).setValidatedEmailToUserByEmailId(IdArgumentCaptor.capture());
+        assertEquals(creatEmailConfirmCode().get().getId(), IdArgumentCaptor.getValue());
+    }
+
+    private Optional<EmailConfirmationCode> creatEmailConfirmCode() {
+        EmailConfirmationCode emailConfirmationCode = new EmailConfirmationCode();
+
+        LocalDateTime localDateTime = LocalDateTime.now();
+        emailConfirmationCode.setExpiryDate(localDateTime.plusDays(1L));
+
+        User user = new User();
+        user.setId(id);
+        user.setEmail("user@mail.ua");
+        emailConfirmationCode.setUser(user);
+
+        return Optional.of(emailConfirmationCode);
+    }
+
+    @ParameterizedTest
+    @MethodSource("testData")
+    void testEmailValidationExceptions(EmailConfirmationCode email, Exception expectedException) {
+        when(emailConfirmationCodeRepository.findById(id)).thenReturn(Optional.ofNullable(email));
+        assertThrows(expectedException.getClass(), () -> sendGridService.validateEmail(id));
+    }
+
+    private static Stream<Arguments> testData() {
```

Comment:
It would be much better to rename the method to a new name that would better describe why those exceptions should be test data